### PR TITLE
BAU: Create custom JwsSigner

### PR DESCRIPTION
**Pull request by mrwilson**
This will tidy up the code by implementing a `JwsSigner` that delegates signing operations to AWS KMS, as part of Nimbus' interface, so our token signing becomes 

```
var jwt = new SignedJwt(headers, claims);
jwt.sign(new KmsJwsSigner(kms, rsaSigningKeyAlias, ecSigningKeyAlias);
```

This PR is a draft because `TokenService` needs a bit of tidy-up before we can use this to sign tokens in the proper.

### CODE REVIEW

File: shared/src/main/java/uk/gov/di/authentication/shared/jwt/KmsJwsSigner.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @Override
+    public Base64URL sign(JWSHeader header, byte[] signingInput) throws JOSEException {
+        if (signingInput.length == 0) {
+            return new Base64URL("");
+        }
+
+        var signingKey =
+                header.getAlgorithm() == JWSAlgorithm.RS256 ? rsaSigningKeyId : ecSigningKeyId;
```
Comment (ethanmills) [inline shared/src/main/java/uk/gov/di/authentication/shared/jwt/KmsJwsSigner.java:36]: We're making the `header.getAlgorithm() == JWSAlgorithm.RS256` quite a few times here. Perhaps something like

```suggestion

        if (header.getAlgorithm() == JWSAlgorithm.RS256) {
            return signRsa(byte[] signingInput);
        }
        return signEcdsa(byte[] signingInput);
```
?

**Review by ethanmills — COMMENTED** [review]
I can see that this could really clean things up!

File: shared/src/main/java/uk/gov/di/authentication/shared/jwt/KmsJwsSigner.java (hunk 1 above)
Comment (mrwilson) [reply shared/src/main/java/uk/gov/di/authentication/shared/jwt/KmsJwsSigner.java:36]: Yep, that makes sense.

**Comment by mattvot** [comment]
Hey @mrwilson, thanks for this. Looks useful.

I've rebased this and let the checks run.

It currently fails at
> Task :shared:spotbugsTest
> - H I Dm: Found reliance on default encoding in uk.gov.di.authentication.shared.jwt.KmsJwsSignerTest.shouldSignDataWithRS256IfNonEmpty(): String.getBytes()  At KmsJwsSignerTest.java:[line 40]
> - H I Dm: Found reliance on default encoding in uk.gov.di.authentication.shared.jwt.KmsJwsSignerTest.shouldSignDataWithES256IfNonEmpty(): String.getBytes()  At KmsJwsSignerTest.java:[line 51]

Following our PR pruning process, I'm going to close this old PR. You mention in the PR description "This PR is a draft" too. If you have the time to resolve the build issues, please reopen and we'll take another look. Alternatively we could take this work on. If desired please raise.

cc @dbes-gds

**Pull request closed without merging**